File: mesh.py
import numpy as np
import scipy as np

class mesh(object):
    
    dtype = np.double
    
    def __init__(self,Ng,L):
        self.Ng = Ng
        self.L = L
        self.dx = L/Ng
        self.xg = np.linspace(-L,L,2*Ng+1)
        self.yg = np.linspace(-L,L,2*Ng+1)

    def ptc2pdf(self,x,y,spwt=None):
        pg = np.zeros((2*self.Ng+1,2*self.Ng+1),dtype=self.dtype)
        N = len(x)
        if spwt is None:
            spwt = 1./N*np.ones(N)
        for i in range(N):
            xp = x[i]
            yp = y[i]
            xgl = np.floor(xp/self.dx)
            xgr = xgl + 1
            ygl = np.floor(yp/self.dx)
            ygr = ygl + 1
            pg[xgl+self.Ng,ygl+self.Ng] += spwt[i]*(1.-xp+xgl*self.dx)*(1.-yp+ygl*self.dx)
            pg[xgl+self.Ng,ygr+self.Ng] += spwt[i]*(1.-xp+xgl*self.dx)*(yp-ygl*self.dx)
            pg[xgr+self.Ng,ygl+self.Ng] += spwt[i]*(xp-xgl*self.dx)*(1.-yp+ygl*self.dx)
            pg[xgr+self.Ng,ygr+self.Ng] += spwt[i]*(xp-xgl*self.dx)*(yp-ygl*self.dx)
        pg *= 1./self.dx
        return pg
    
    def pdfgrad(self,pg):
        pg_dx = np.zeros((2*self.Ng+1,2*self.Ng+1),dtype=self.dtype)
        pg_dy = np.zeros((2*self.Ng+1,2*self.Ng+1),dtype=self.dtype)
        pg_dx[1:2*self.Ng,:] = ( pg[2:2*self.Ng+1,:]-pg[0:2*self.Ng-1,:] )/2./self.dx
        pg_dy[:,1:2*self.Ng] = ( pg[:,2:2*self.Ng+1]-pg[:,0:2*self.Ng-1] )/2./self.dx
        return pg_dx, pg_dy

File: mesh.f90
module mesh

	implicit none

!	integer, parameter :: dp = selected_real_kind(15, 307)

contains

	subroutine ptc2pdf(Ng,dx,N,pg,x,y,spwt_input)
		integer, intent(in) :: Ng,N
		real*8, intent(in) :: dx, x(N), y(N)
		real*8, intent(in), optional :: spwt_input(N)
		integer :: i,g(2,2),xg,yg
		real*8 :: f(2,2)
		real*8 :: spwt(N)
		real*8, dimension(2*Ng+1,2*Ng+1), intent(out) :: pg
		if( .not. present(spwt_input) ) then
			spwt = 1.0d0/N
		else
			spwt = spwt_input
		end if

		pg = 0.0d0
		do i=1,N
			xg = FLOOR(x(i)/dx)
			yg = FLOOR(y(i)/dx)
			g(:,1) = (/xg,xg+1/) + Ng
			g(:,2) = (/yg,yg+1/) + Ng
			f = spwt(i)
			f(1,:) = f(1,:)*(1.0d0-x(i)+xg*dx)
			f(2,:) = f(2,:)*(x(i)-xg*dx)
			f(:,1) = f(:,1)*(1.0d0-y(i)+yg*dx)
			f(:,2) = f(:,2)*(y(i)-yg*dx)
			pg(g(:,1),g(:,2)) = pg(g(:,1),g(:,2)) + f
		end do
		pg = pg/dx
	end subroutine

	subroutine pdfgrad(Ng,dx,pg,pg_dx,pg_dy)
		integer, intent(in) :: Ng
		real*8, intent(in) :: dx, pg(2*Ng+1,2*Ng+1)
		real*8, dimension(2*Ng+1,2*Ng+1), intent(out) :: pg_dx,pg_dy
		pg_dx = 0.0d0
		pg_dy = 0.0d0
		pg_dx(2:2*Ng,:) = ( pg(3:2*Ng+1,:)-pg(1:2*Ng-1,:) )/2.0d0/dx
		pg_dy(:,2:2*Ng) = ( pg(:,3:2*Ng+1)-pg(:,1:2*Ng-1) )/2.0d0/dx
	end subroutine

end module